pkg/deviceshifu/deviceshifuhttp/deviceshifuhttp_test.go

```diff
@@ -186,3 +194,256 @@ func TestCreatehttpURIStringNoQuery(t *testing.T) {
+			default:
+				w.Header().Add("Content-Type", "application/json")
+				w.WriteHeader(http.StatusOK)
+				println("ds default request, path:", path)
+			}
+		} else {
+			println("invalid method")
+		}
+	}))
+	return server
+}
+
+func mockHandlerServer() *httptest.Server {
+	// catch handler http request and response properly with specific paths
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		switch path {
+		case "/test_instruction":
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			println("handler get the instruction and executed.")
+		default:
+			w.WriteHeader(http.StatusOK)
+			println("hs get default request, path:", path)
+		}
+
+	}))
+	return server
+}
+
+func mockDeviceShifuInstruction() *deviceshifubase.DeviceShifuInstruction {
+	return &deviceshifubase.DeviceShifuInstruction{
+		DeviceShifuInstructionProperties: []deviceshifubase.DeviceShifuInstructionProperty{
+			{
+				ValueType:    "testing",
+				ReadWrite:    "rw",
+				DefaultValue: "0",
+			},
+		},
+		DeviceShifuProtocolProperties: map[string]string{
+			"test_key": "test_value",
+		},
+	}
+}
+
+func Test_collectHTTPTelemtries(t *testing.T) {
+	ts := mockTelemetryServer()
+	addr := strings.Split(ts.URL, "//")[1]
+	mockDevice := &DeviceShifuHTTP{
+		base: &deviceshifubase.DeviceShifuBase{
+			Name: "test",
+			EdgeDevice: &v1alpha1.EdgeDevice{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "test_namespace",
+				},
+				Spec: v1alpha1.EdgeDeviceSpec{
+					Address:  &addr,
+					Protocol: (*v1alpha1.Protocol)(strPointer(string(v1alpha1.ProtocolHTTP))),
+				},
+			},
+			DeviceShifuConfig: &deviceshifubase.DeviceShifuConfig{
+				Telemetries: &deviceshifubase.DeviceShifuTelemetries{
+					DeviceShifuTelemetrySettings: &deviceshifubase.DeviceShifuTelemetrySettings{
+						DeviceShifuTelemetryTimeoutInMilliseconds:    int64Pointer(10),
+						DeviceShifuTelemetryDefaultPushToServer:      boolPointer(true),
+						DeviceShifuTelemetryDefaultCollectionService: strPointer("test_endpoint-1"),
+					},
+					DeviceShifuTelemetries: map[string]*deviceshifubase.DeviceShifuTelemetry{
+						"device_healthy": {
+							DeviceShifuTelemetryProperties: deviceshifubase.DeviceShifuTelemetryProperties{
+								DeviceInstructionName: strPointer("mock_testing"),
+								PushSettings: &deviceshifubase.DeviceShifuTelemetryPushSettings{
+									DeviceShifuTelemetryPushToServer:      boolPointer(false),
+									DeviceShifuTelemetryCollectionService: strPointer("test_endpoint-1"),
+								},
+								InitialDelayMs: intPointer(1),
+							},
+						},
+					},
+				},
+			},
+			RestClient: mockRestClientFor(addr, "{\"spec\": {\"address\": \"http://192.168.15.48:12345/test_endpoint-1\",\"type\": \"HTTP\"}}", t),
+		},
+	}
+
+	res, err := mockDevice.collectHTTPTelemtries()
+	assert.Equal(t, true, res)
+	assert.Nil(t, err)
+
+}
+
+func boolPointer(b bool) *bool {
```

Comment:
You already had those helper functions in https://github.com/Edgenesis/shifu/blob/main/pkg/deviceshifu/deviceshifubase/push_telemetry_test.go#L17 
Put them into some common place, try to avoid copying them twice.